src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexParser.cs

```diff
@@ -513,33 +516,25 @@ private RegexNode ScanReplacement()
         {
             _concatenation = new RegexNode(RegexNodeKind.Concatenate, _options);
 
-            while (true)
+            while (_pos < _pattern.Length)
             {
-                int c = _pattern.Length - _pos;
-                if (c == 0)
-                {
-                    break;
-                }
-
                 int startpos = _pos;
 
-                while (c > 0 && _pattern[_pos] != '$')
-                {
-                    _pos++;
-                    c--;
-                }
+                _pos = _pattern.IndexOf('$', _pos);
+                if (_pos == -1)
+                    _pos = _pattern.Length;
 
-                AddConcatenate(startpos, _pos - startpos, isReplacement: true);
+                AddToConcatenate(startpos, _pos - startpos, isReplacement: true);
 
-                if (c > 0)
+                if (_pos < _pattern.Length)
                 {
                     if (_pattern[_pos++] == '$')
```

Comment:
In what situation will `_pattern[_pos]` here _not_ equal '$'?  Does AddToConcatenate change _pos? If not, aren't we guaranteed that _pattern[_pos] == '$' since if _pos < _pattern.Length that means IndexOf found the '$'?